wurst/event/DamageModify.wurst

```diff
@@ -0,0 +1,462 @@
+	Features custom instance lifetime.
+	Callbacks have Damage instance as an argument.
+
+	Damage instance can be created before damage invokation, thus it can be modified anyhow before the event.
+	It is guaranteed to be passed to modifiers and listeners during the event invoked through `.deal()` method.
+	And then the instance is still accessable (in case it was saved), allowing user to easily retrieve event data.
+
+	System is guaranteed to work as in documentation only if there's no listeners in DamageDetection after this package's one.
+	The system does not handle 0 damage events.
+	
+	Has several cases of possibly unexpected behavior:
+	- Actual amount can not go below zero or above target's HP.
+	- Event Damage is set to .actualAmount, not .amount.
+	- Damage instance can not be modified from inside listeners.
+
+	There will be output to debug log stream if anything would go wrong,
+	for example if system will fail to pass/catch Damage instance in between
+	damage instance creation and it's handling (.deal() and .handle()).
+	
+	Examples:
+	
+	// Leech implementation #1
+	function onLeechEffect1()
+		// Using wrapper from the package, event data is retrieved easily.
+		let d = GetTriggerUnit().damageTarget(GetSpellTargetUnit(), LEECH_DAMAGE, ATTACK_TYPE_NORMAL, 0)
+
+		d.source.addHP(d.getActual()) // Heal is capped at target's hp
+		d.source.addEffect(Abilities.vampiricAuraTarget, "origin").destr()
+
+		d.releaseInstance() // Don't forget to free the instance
+
+	init
+		registerSpellEffectEvent(LEECH_ID, function onLeechEffect1)
+
+	// Leech implementation #2
+	function onLeechEffect2()
+		GetTriggerUnit().damageTarget(GetSpellTargetUnit(), LEECH_DAMAGE, ATTACK_TYPE_NORMAL, LEECH_ID)
+
+	init
+		registerSpellEffectEvent(LEECH_ID, function onLeechEffect2)
+		addDamageListener() d ->
+			if d.data == LEECH_ID
+				d.source.addHP(d.getActual())
+				d.source.addEffect(Abilities.vampiricAuraTarget, "origin").destr()
+
+	// Simple damage resistance buff
+	import UniqueDamageListening
+
+	function onBuffEffect()
+		let target = GetSpellTargetUnit()
+		let effect = target.addEffect(Abilities.voodooAuraTarget, "origin")
+		let cb = addTargetDamageModifier() d ->
+			d.amount -= d.initialAmount * .5
+		doAfter(BUFF_DURATION) ->
+			removeTargetDamageModifier(cb)
+			effect.destr()
+
+	init
+		registerSpellEffectEvent(BUFF_ID, function onLeechEffect2)
+*/
+
+// Configuration
+
+/** If Damage instance wasn't created by one of DamageModify package's functions
+	(or `new Damage(...)`), then Damage.data would be set to this value */
+@configurable public constant DAMAGE_NONTRIGGERED_DATA = -1
+
+/** Maximum amount of damage events that can happen in a single event */
+@configurable public constant RECURSION_TRESHOLD = 32
+
+/** Describes default way to create Damage instance if damage event doesn't
+	correspond to any Damage instance. */
+@configurable function getDamageFromEvent() returns Damage
```

Comment:
This creates new Damage instance, that's why it is different from Blizzard notation. It's not critical since it's a private function, but I would want to hear what @Frotty thinks about it too before changing it.

Reply:
Why is it configurable?
I would probably choose something like `getEventDamageInstance` or so.

Reply:
Sounds better for me.
Configurable because it includes sorting out 0 damage and some default values like `attack type chaos` in case of spell damage. Or someone would like to replace dtype, or to remove non-triggered damage entirely.